### PR TITLE
Create a flight-python build

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
Python version `3.8.10` has been packaged as `flight-python`.

The `flight-python-system-*` version has been repurposed to denote the `MAJOR.MINOR` release version. Because python packages are typically packaged with `penv`, they are coupled to a particular minor release. Using the system version is a shortcut for specifying these bounds.

The `flight-cert` package has been rebuilt to use `flight-python`.

### CODE REVIEW

File: builders/flight-certbot/config/software/enforce-flight-python.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+license :project_license
+skip_transitive_dependency_licensing true
+
+build do
+  raise "Flight Python is not installed!" if ! File.exists?('/opt/flight/bin/python3')
+  python_system = `/opt/flight/bin/python3 --version`.chomp.split(' ').last.gsub(/\.\d+\Z/, '')
+  unless python_system == version
+    raise "Flight Python has the incorrect system version: #{python_system} (expected #{version})"
+  end
```
Comment (benarmston) [inline builders/flight-certbot/config/software/enforce-flight-python.rb:42]: What's the purpose of this check?  I understand that we do this for `flight-runway` (i.e., ruby), but in that case we're using ruby (and bundler) as part of the build chain.  Are we using `python` as part of the build chain?  Is there some other reason for this?

File: builders/flight-certbot/config/software/flight-certbot.rb [hunk 2]
```diff
@@ -29,15 +29,16 @@
...
-
 license 'Apache-2.0'
 license_file 'LICENSE.txt'
 skip_transitive_dependency_licensing true
 
 build do
   env = with_embedded_path("PIPENV_VENV_IN_PROJECT" => 'true')
 
+  # Place the openflight version of python onto the path
+  env['PATH'] = "/opt/flight/bin:#{env['PATH']}"
```
Comment (benarmston) [inline builders/flight-certbot/config/software/flight-certbot.rb:40]: I needed the `python` from `flight-python` to fix an issue with `flight env create easybuild`.  I decided there that adding `/opt/flight/bin/` to the path brought in too much.  I've added a commit so that we have

* `/opt/flight/opt/python/bin/{python,python3,pip,pip3}`
* `/opt/flight/bin/{python,python3,pip,pip3}`

This should be updated to add `/opt/flight/opt/python/bin/`.

File: builders/flight-python/config/projects/flight-python.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+# These files are installed as part of the flight-python software build.  They
+# are symlinks to `/opt/flight/opt/python/bin/`.
+#
+# Unfortunately, it appears that they can't be included in the package from
+# that software.  So they get included as extra_package_files which
+# unfortunately adds them to the package as regular files not symlinks.
+%w(python3 python pip3 pip).each do |f|
+  extra_package_file "/opt/flight/bin/#{f}"
+end
```
Comment (benarmston) [inline builders/flight-python/config/projects/flight-python.rb:59]: If you know of a better workaround than this I'd love to hear it.

Currently, the build machine has regular files at `/opt/flight/opt/python/bin/{python3,pip3}` and symlinks at `/opt/flight/opt/python/bin/{python,pip}` and `/opt/flight/bin/{python3,pip3,python,pip}`.

The RPM gets regular files at `/opt/flight/opt/python/bin/{python3,pip3}`  and `/opt/flight/bin/{python3,pip3,python,pip}` and symlinks at `/opt/flight/opt/python/bin/{python,pip}`.

File: builders/flight-python/config/software/flight-python.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+build do
+  # Check that flight-python has not been installed or previously built.
+  # NOTE: this is done when the software file is loaded
+
+  if %w(python3 python pip3 pip).any? {|f| File.exists?(File.join("/opt/flight/bin", f))}
+    raise <<~ERROR
+      flight-python can not be built when existing version has been installed!
+      Please remove the system version before continuing
+    ERROR
+  end
```
Comment (benarmston) [inline builders/flight-python/config/software/flight-python.rb:17]: What are we trying to prevent here?

File: builders/flight-python/config/software/python.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with this project. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on OpenFlight Omnibus Builder, please visit:
+# https://github.com/openflighthpc/openflight-omnibus-builder
+#===============================================================================
+name 'python'
```
Comment (benarmston) [inline builders/flight-python/config/software/python.rb:27]: Is there a reason that you didn't reuse the software definition at `share/python3`?

File: builders/flight-python/config/software/readline.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with this project. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on OpenFlight Omnibus Builder, please visit:
+# https://github.com/openflighthpc/openflight-omnibus-builder
+#===============================================================================
+name "readline"
```
Comment (benarmston) [inline builders/flight-python/config/software/readline.rb:27]: Is there a reason that you didn't reuse the software definition at `share/readline`?

File: builders/flight-python/config/software/sqlite.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with this project. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on OpenFlight Omnibus Builder, please visit:
+# https://github.com/openflighthpc/openflight-omnibus-builder
+#===============================================================================
+name 'sqlite'
```
Comment (benarmston) [inline builders/flight-python/config/software/sqlite.rb:27]: Is there a reason that you didn't reuse the software definition at `share/sqlite`?

File: builders/flight-python/dist/bin/pip3 [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    shopt -s nullglob
+    for a in "${flight_ROOT}"/libexec/hooks/*.sh; do
+      source "${a}"
+    done
+    shopt -u nullglob
+  fi
+}
+
+flight_ROOT=${flight_ROOT:-$(cd $(dirname ${BASH_SOURCE[0]})/.. && pwd)}
+_setup
```
Comment (benarmston) [inline builders/flight-python/dist/bin/pip3:47]: I can't decide if I think this is correct or not.  I'm not sure that we want to write our own tools in `python`, but we do want to be able to rely on `python` for our tool's dependencies.  Did you have something particular in mind when you added this `_setup`?

**Review by benarmston — CHANGES_REQUESTED** [review]
I've added a commit that changes where the executables are installed.  I wanted to be able to add just the `flight-python` executables to `PATH`.

I've also attempted to package the symlinks in the RPM, with some success.  I'm somewhat concerned about leaving symlinks to executables when the RPM is removed.

I've also separated out the `flight-python` software definition from the `python` software definition.  This is in better keeping with how some of our other packages work and allows fiddling with `flight-python` without destroying the `python` build cache.

With those changes made, I've got a few questions for you.  I've added them inline below.

Also, the package doesn't build on either Centos8 or Ubuntu2004 (with or without my changes) can you take a look at those build errors too please?

File: builders/flight-python/config/software/flight-python.rb (hunk 4 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/flight-python.rb:17]: You can't build `flight-python` if an existing version is installed within `/opt/flight/opt/python`. Doing so can create conflicts between the newly built files and the installed `rpm`.

File: builders/flight-python/config/software/python.rb (hunk 5 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/python.rb:27]: Mainly because I forgot about the shared version.

However moving forward we won't need the `share/python3` as we can specify a `runtime_dependency: flight-python`. It can be removed once all the relevant package builders have been updated.

File: builders/flight-python/config/software/readline.rb (hunk 6 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/readline.rb:27]: Ditto,

However we can probably use the shared version.

File: builders/flight-python/config/software/sqlite.rb (hunk 7 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/sqlite.rb:27]: Ditto,

However I don't think anything uses `sqlite` apart from `python`. So the shared version *may* be removed once the packages are ported.

File: builders/flight-python/dist/bin/pip3 (hunk 8 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/dist/bin/pip3:47]: I'm not  exactly sure what you are referring to.

I cargo programmed this from `/o/f/b/yarn`:

```
_setup() {
  local a xdg_config
  IFS=: read -a xdg_config <<< "${XDG_CONFIG_HOME:-$HOME/.config}:${XDG_CONFIG_DIRS:-/etc/xdg}"
  for a in "${xdg_config[@]}"; do
    if [ -e "${a}"/flight.rc ]; then
      source "${a}"/flight.rc
      break
    fi
  done
  if [ -d "${flight_ROOT}"/libexec/hooks ]; then
    shopt -s nullglob
    for a in "${flight_ROOT}"/libexec/hooks/*.sh; do
      source "${a}"
    done
    shopt -u nullglob
  fi
}

flight_ROOT=${flight_ROOT:-$(cd $(dirname ${BASH_SOURCE[0]})/.. && pwd)}
_setup
unset _setup

exec ${flight_ROOT}/bin/node \
     ${flight_ROOT}/opt/nodejs/bin/yarn.js "$@"

```

File: builders/flight-certbot/config/software/enforce-flight-python.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply builders/flight-certbot/config/software/enforce-flight-python.rb:42]: A few packages require `pip` as part of the build (e.g. `flight-certbot`). The builds are dependent on the `MAJOR.MINOR` version of `python`.

File: builders/flight-python/config/software/flight-python.rb (hunk 4 above)
Comment (benarmston) [reply builders/flight-python/config/software/flight-python.rb:17]: This also catches the case that `flight-python` has simply been built previously, which is very annoying.  Perhaps I introduced that with my changes to where the binaries are being installed.  If so, that might be relatively simple to undo.

Do you recall if this check caused problems with rebuilds previously?

File: builders/flight-python/config/software/python.rb (hunk 5 above)
Comment (benarmston) [reply builders/flight-python/config/software/python.rb:27]: In that case.  There are some differences between the two software definitions.  Can you take a look and convince yourself (and me) that we want those differences.

File: builders/flight-python/config/software/readline.rb (hunk 6 above)
Comment (benarmston) [reply builders/flight-python/config/software/readline.rb:27]: Using the shared version sounds good.  Again there are differences.  Just the version number.

File: builders/flight-python/config/software/sqlite.rb (hunk 7 above)
Comment (benarmston) [reply builders/flight-python/config/software/sqlite.rb:27]: Big differences between the two this time.  Can you take a look at them and convince yourself (and me) that we want those differences.

File: builders/flight-python/dist/bin/pip3 (hunk 8 above)
Comment (benarmston) [reply builders/flight-python/dist/bin/pip3:47]: It's correct for `node`, because we have APIs written in JavaScript that we want to start in the correct environment.  I guess we have some tools written in python already `certbot` and `directory`.  With `certbot`, we always use it via a wrapper, `flight www cert-gen`, and that wrapper sets up the flight environment.  I'm not sure what we do with directory.

That's mostly thinking out loud.

This is fine.  Let's keep it.

File: builders/flight-certbot/config/software/enforce-flight-python.rb (hunk 1 above)
Comment (benarmston) [reply builders/flight-certbot/config/software/enforce-flight-python.rb:42]: Fair enough.  Sounds like something we should be doing then.

File: builders/flight-python/config/software/flight-python.rb (hunk 4 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/flight-python.rb:17]: Hmm it shouldn't be picking up old builds?

The `/o/f/bin/python` files are `extra_package_files`, IIRC these should only be present when installing the `rpm`. I may be wrong, but this should not happen at build time.

There maybe something funky happening with symlinks?

Comment (benarmston) [reply builders/flight-python/config/software/flight-python.rb:17]: It's likely a hold over from my attempt to make `/o/f/b/python` et al symlnks.  https://github.com/openflighthpc/openflight-omnibus-builder/pull/130/commits/7236eed068585e6bc803e75f1e0baabcef88fe46#diff-092b957480f07bb0a69e66d1054fb0f94d70fed2ca3c4be2461fa6fb244cedc1R23.  The block of code there creates symlinks on the build machine at `/opt/flight/bin/python` et al.  As symlink packaging isn't working correctly, perhaps changing that to copy `dist/bin/*` to `opt/flight/bin/` is the better solution.

Comment (benarmston) [reply builders/flight-python/config/software/flight-python.rb:17]: It no longer fails on reinstall. I decided on abandoning symlinks though.

Comment (benarmston) [reply builders/flight-python/config/software/flight-python.rb:17]: > You can't build `flight-python` if an existing version is installed within `/opt/flight/opt/python`. Doing so can create conflicts between the newly built files and the installed `rpm`.

What kind of conflicts are you referring to here?

Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/flight-python.rb:17]: Every file within `/opt/flight/opt/python` creates a conflict. We build all packages in place (directly within `/opt/flight/opt/<app>`) so that anything that hard codes a path will be correct.

This isn't an issue for the majority of builds (e.g, `flight-job`) as it will not be installed on the `builder`. This means `/opt/flight/opt/job` will be an empty directory ready for the build. It also means it is safe to run `rm -rf` or `omnibus clean --purge` over the directory before building.

The same does not apply `flight-python` as it is required to build other packages (e.g. `flight-cerbot`). This means `/opt/flight/opt/python` may not be empty and `rm -rf`/ `omnibus clean --purge` are now dangerous operations.

---

This is why I have added a check to ensure `flight-python` is not installed before building `flight-python`. Technically `flight-runway` (maybe `flight-nodejs`?) is susceptible to the same issue. However these have legacy builders before all the idiosyncrasies where identified.

File: builders/flight-python/config/software/python.rb (hunk 5 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/python.rb:27]: So the `env` handling is a little bit different, I think this is primarily due to cargo programming from `omnibus-software` upstream definitions.

There seems to be the "standard" methods `with_standard_compiler_flags(with_embedded_path())` that sets the compiler flags vs. setting them manually. AFAICS there is no particularly advantage of one over the other.

I am partially to using the standard methods, mainly because its a bit more verbose in what it sets.

---

Apart from that, there where a couple of missing dependencies. I have added `gdbm` back, which should fix your build issues on ubuntu.

I've also added `bzip2` back, but I don't think it's actually being used.

File: builders/flight-python/config/software/readline.rb (hunk 6 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/readline.rb:27]: So there are a couple of difference:

The license in the original file looks wrong. `SQLITE` uses a "blessing" that puts it in the open domain. The original `libxslt` license looks wrong.

A latter version of `sqlite` is being built, `3.35.5` vs `3.31.1`. Don't see this causing an issue, it will almost certainly be compatible. We should probably look to build with the latter version.

The `relative_path` is a bit different, but this is a non-issue. It gives the name of the directory after the tarball is extracted. Don't know why it is different, possible due to a latter version? As long as the package builds, it is fine.

We aren't using the `--disable-readline` flag in the build. We are building it with `readline`, so no point disabling it.

Finally we are using the standard `env = with_standard_compiler_flags(with_embedded_path())` instead of the custom configuration. Generally I am partially to this method for consistency.

---

tl;dr There are a few difference, but they shouldn't be an issue.

File: builders/flight-python/config/software/sqlite.rb (hunk 7 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/sqlite.rb:27]: So there are a couple of difference:

The license in the original file looks wrong. `SQLite` uses a "blessing" that puts it in the public domain. The original `libxslt` license looks wrong.

A latter version of `sqlite` is being built, `3.35.5` vs `3.31.1`. Don't see this causing an issue, it will almost certainly be compatible. We should probably look to build with the latter version.

The `relative_path` is a bit different, but this is a non-issue. It gives the name of the directory after the tarball is extracted. Don't know why it is different, possible due to a latter version? As long as the package builds, it is fine.

~We aren't using the `--disable-readline` flag in the build. We are building it with `readline`, so no point disabling it.~

Finally we are using the standard `env = with_standard_compiler_flags(with_embedded_path())` instead of the custom configuration. Generally I am partially to this method for consistency.

---

tl;dr There are a few difference, but they shouldn't be an issue.

File: builders/flight-python/config/software/readline.rb (hunk 6 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/readline.rb:27]: Removed

File: builders/flight-python/config/software/sqlite.rb (hunk 7 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/sqlite.rb:27]: Nope: We need `--disabled-readline` because `sqlite` uses a different version on `centos8`

File: builders/flight-python/dist/bin/pip3 (hunk 8 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/dist/bin/pip3]: :+1:

File: builders/flight-certbot/config/software/flight-certbot.rb (hunk 2 above)
Comment (benarmston) [reply builders/flight-certbot/config/software/flight-certbot.rb:40]: @WilliamMcCumstie can we get this updated please.

File: builders/flight-python/config/software/flight-python.rb (hunk 4 above)
Comment (benarmston) [reply builders/flight-python/config/software/flight-python.rb:17]: I'm still not sure what you mean by conflict.  RPMs can conflict with each other in a couple of ways.  The RPM itself can be flagged to `Conflicts: other-rpm`; or an RPM can own a file that is also owned by another RPM.  We have neither of those situations here.

The installed `flight-python` RPM is the same RPM as the one being built, albeit a different version.  An RPM can't `Conflicts` with itself, even a later version.  A file ownership conflict cannot exist between two different versions of an RPM.

I'm left confused as to what kind of conflict you are referring to.

I understand that `yum install flight-python` will result in files being placed into `/opt/flight/bin/` and `/opt/flight/opt/python`.  I also understand that some of those files may be overwritten when running `omnibus build flight-python`.

What I don't understand is what issue that is going to cause.

I don't understand why `omnibus clean --purge` is "dangerous".  I can see that it could break an installation of the `flight-python` RPM package.  But that is easily recoverable, by reinstalling the RPM.

This mechanism also recovers from the issue of some files being overwritten when running `omnibus build flight-python`.

---

Regarding `flight-runway` and `flight-nodejs`, it may be that they also suffer from this issue and should be updated.  But before I can make a decision on that I need to understand what this issue is.

File: builders/flight-python/config/software/sqlite.rb (hunk 7 above)
Comment (benarmston) [reply builders/flight-python/config/software/sqlite.rb:27]: Does `env = with_standard_compiler_flags(with_embedded_path())` produce the same environment as is set manually?

This should be renamed to `sqlite3`.  I don't think it is a good idea to introduce arbitrary naming variations from upstream, and I can't see a reason why we'd want to deviate from the name selected by upstream.

File: builders/flight-python/config/software/python.rb [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  #
+  # The './configure --help' output contains various option flags to control the
+  # build. The --with-libs options *looks* to be the appropriate flag to force
+  # python to compile with zlib and openssl, however it './configure'
+  # to error (for reasons ¯\_(ツ)_/¯):
+  #
+  # ./configure --with-libs='libz'
+  # ...
+  # Fatal: You must get working getaddrinfo() function.
+  #        or you can specify "--disable-ipv6".
```
Comment (benarmston) [inline builders/flight-python/config/software/python.rb:66]: Is this comment still relevant?  Are we compiling python with `zlib` and `openssl`?

File: builders/flight-python/config/software/flight-python.rb (hunk 4 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/flight-python.rb:17]: I think we are using two different meanings of the word "conflict". I mean it in a generic sense when two files try and write to the same location.

> The RPM itself can be flagged to `Conflicts: other-rpm`; or an RPM can own a file that is also owned by another RPM. We have neither of those situations here.

I'm not referring to multiple `rpms` conflicting with each other. I am referring to the `builder` conflicting with an `rpm`. As `yum` is not involved in this, the `Conflict:` flag does not work.

> The installed flight-python RPM is the same RPM as the one being built, albeit a different version

That's almost always not true as they will be different versions. Best case scenario they will have the same filenames but updated versions. In this case, all the old versions _should_ be replaced (hopefully, maybe?). However there are multiple pit falls including but not limited to:

* We don't know how `make install` works, it may not replace the file, and
* If the filenames change, the old stale versions will be included in the new `rpm`.

Overall, there are too many *shoulds* and *maybes* for my liking. We definitely want to be building from a clean slate hence the check.

> I don't understand why `omnibus clean --purge` is "dangerous". I can see that it could break an installation of the flight-python RPM package. But that is easily recoverable, by reinstalling the RPM.

Yes this is technically true, but I would always classify breaking a system install as "dangerous". We should avoid doing so if possible.

Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/flight-python.rb:17]: > This mechanism also recovers from the issue of some files being overwritten when running omnibus build flight-python.

It doesn't recover if the filenames have changed. This can occur quite frequently as `python3` hard codes the `MAJOR.MINOR` release into some of its paths:

```
$ ls  /opt/flight/opt/python/embedded/include/python3.8/
abstract.h         cpython                import.h               odictobject.h      pyhash.h       setobject.h
asdl.h             datetime.h             internal               opcode.h           pylifecycle.h  sliceobject.h
ast.h              descrobject.h          interpreteridobject.h  osdefs.h           pymacconfig.h  structmember.h
bitset.h           dictobject.h           intrcheck.h            osmodule.h         pymacro.h      structseq.h
bltinmodule.h      dtoa.h                 iterobject.h           parsetok.h         pymath.h       symtable.h
boolobject.h       dynamic_annotations.h  listobject.h           patchlevel.h       pymem.h        sysmodule.h
bytearrayobject.h  enumobject.h           longintrepr.h          picklebufobject.h  pyport.h       token.h
bytes_methods.h    errcode.h              longobject.h           pyarena.h          pystate.h      traceback.h
bytesobject.h      eval.h                 marshal.h              pycapsule.h        pystrcmp.h     tracemalloc.h
cellobject.h       fileobject.h           memoryobject.h         pyconfig.h         pystrhex.h     tupleobject.h
ceval.h            fileutils.h            methodobject.h         pyctype.h          pystrtod.h     typeslots.h
classobject.h      floatobject.h          modsupport.h           py_curses.h        Python-ast.h   ucnhash.h
codecs.h           frameobject.h          moduleobject.h         pydebug.h          Python.h       unicodeobject.h
code.h             funcobject.h           namespaceobject.h      pydtrace.h         pythonrun.h    warnings.h
compile.h          genobject.h            node.h                 pyerrors.h         pythread.h     weakrefobject.h
complexobject.h    graminit.h             object.h               pyexpat.h          pytime.h
context.h          grammar.h              objimpl.h              pyfpe.h            rangeobject
```

File: builders/flight-python/config/software/sqlite.rb (hunk 7 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/sqlite.rb]: Renamed: https://github.com/openflighthpc/openflight-omnibus-builder/pull/130/commits/d4b7861806c761dfaf7b1e9c4bafc44b72e7c145

File: builders/flight-python/config/software/python.rb (hunk 9 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/python.rb:66]: Yep that is still relevant

We are compiling with `zlib` and `openssl` but only because `zlib-devel` and `openssl-devel` is installed on the system.

`python3` does not appear to have options to force the compiler to include the libraries or error.

File: builders/flight-python/config/software/sqlite.rb (hunk 7 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/sqlite.rb]: So the standard flags are the following:

```
CFLAGS="-I/opt/flight/opt/python/embedded/include -O2 -D_FORTIFY_SOURCE=2 -fstack-protector"
CPPFLAGS="-I/opt/flight/opt/python/embedded/include -O2 -D_FORTIFY_SOURCE=2 -fstack-protector"
CXXFLAGS="-I/opt/flight/opt/python/embedded/include -O2 -D_FORTIFY_SOURCE=2 -fstack-protector"
LDFLAGS="-Wl,-rpath,/opt/flight/opt/python/embedded/lib -L/opt/flight/opt/python/embedded/lib"
LD_RUN_PATH="/opt/flight/opt/python/embedded/lib"          
OMNIBUS_INSTALL_DIR="/opt/flight/opt/python"               
PATH="/opt/flight/opt/python/bin:/opt/flight/opt/python/embedded/bin:/home/vagrant/openflight-omnibus-builder/vendor/ruby/2.7.0/bin:/usr/local/rvm/gems/ruby-2.7.2/bin:/usr/local/
rvm/gems/ruby-2.7.2@global/bin:/usr/local/rvm/rubies/ruby-2.7.2/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/usr/local/rvm/bin:/home/vagrant/.local/bin:/home/vagrant/bin"
PKG_CONFIG_PATH="/opt/flight/opt/python/embedded/lib/pkgconfig"
```

Obviously there are additional environment variables (mostly concerning `c++`). These will be ignored by `gcc`, and there is already precedence for doing this. 

The three original env vars where:

```
"LDFLAGS" => "-Wl,-rpath,#{install_dir}/embedded/lib -L#{install_dir}/embedded/lib -I#{install_dir}/embedded/include",
"CFLAGS" => "-L#{install_dir}/embedded/lib -I#{install_dir}/embedded/include",
"LD_RUN_PATH" => "#{install_dir}/embedded/lib"
```

Of which, `LD_RUN_PATH` is the same.

The `CFLAGS` and `LDFLAGS` have a minor difference. It looks like the original version of `CFLAGS` includes a `-L` directive to the linker and `LDFLAGS` a `-I` directive to the compiler. This doesn't really make sense, my guess the flags where duplicated at some point. I am comfortable that `with_standard_env` omits the duplicate directives where appropriate.

Otherwise the `with_standard_env` has the following additional options:

* `-fstack-protector`: Additional protections against buffer overflows,
* `-D_FORTIFY_SOURCED=2`: Run additional compiler time checks on some `libc` functions (when supported),
* `-O2`: Apply a bunch of compiler optimisations

Overall `with_standard_env` looks to be what we want

File: builders/flight-certbot/config/software/flight-certbot.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply builders/flight-certbot/config/software/flight-certbot.rb]: So we don't need both on this `PATH`; one or the other is sufficient. I've gone with the internal `/opt/flight/opt/python/bin`.

However the original shims where broken as they directly `cp` the distro versions. This meant `flight_ROOT` was being set to `/opt/flight/opt/python` and breaks everything.

Instead I've updated the shims to use an `exec`.

Comment (WilliamMcCumstie) [reply builders/flight-certbot/config/software/flight-certbot.rb]: 36f8b33afde87cd8fe47d7b028ba904a02c48a29

Comment (WilliamMcCumstie) [reply builders/flight-certbot/config/software/flight-certbot.rb]: Changed my mind, I got `symlinks` working in `/opt/flight/opt/python/bin`:

https://github.com/openflighthpc/openflight-omnibus-builder/pull/130/commits/a8b9b914ff401191af033be3ea721aa6c0ae58d4

File: builders/flight-python/config/projects/flight-python.rb (hunk 3 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/projects/flight-python.rb]: The trick is to create the symlinks within `/opt/flight/opt/python/bin` within the software file. This way `omnibus` will handle the symlink natively.

https://github.com/openflighthpc/openflight-omnibus-builder/pull/130/commits/a8b9b914ff401191af033be3ea721aa6c0ae58d4

Comment (WilliamMcCumstie) [reply builders/flight-python/config/projects/flight-python.rb]: As far as `/opt/flight/bin` goes, they have to be regular files

File: builders/flight-certbot/config/software/flight-certbot.rb (hunk 2 above)
Comment (benarmston) [reply builders/flight-certbot/config/software/flight-certbot.rb]: > So we don't need both on this `PATH`; one or the other is sufficient. I've gone with the internal `/opt/flight/opt/python/bin`.

Good.  That's what I asked for :-)

File: builders/flight-python/config/software/flight-python.rb (hunk 4 above)
Comment (benarmston) [reply builders/flight-python/config/software/flight-python.rb:73]: > I think we are using two different meanings of the word "conflict". I mean it in a generic sense when two files try and write to the same location.

I had been trying to avoid using "conflict" for exactly this reason; it isn't well-defined in this scenario.  Hence my requests for you to define it.

I think I now understand your concerns.

1. When building an RPM, the contents of `/opt/flight/opt/<NAME>` and some of the contents of `/opt/flight/{bin,libexec,...}` are included in the built RPM.
2. We want to ensure that artefacts of previous builds are not included in the newly built RPM.
3. For many of our builders we do not expect any of the built RPMs to be installed on the build machine. So we can achieve (2) by running `clean --purge`.
4. For some of our builders we do expect the built RPM to be installed on the build machine.  Running `clean --purge` will remove files belonging to that installed RPM.  Running `build` will overwrite/replace files belonging to that installed RPM.  This can result in an unpredictable mix of files belonging to the installed RPM and/or the newly built RPM.
5. Using that unpredicatble mix to build additional packages is likely to have unpredictable and undesirable outcomes.

Could you clarify that I've understood you correctly.

The above certainly seems like a valid concern.

File: builders/flight-python/config/software/sqlite.rb (hunk 7 above)
Comment (benarmston) [reply builders/flight-python/config/software/sqlite.rb]: Yep.  Those look fine to me.  Thanks for checking that.

File: builders/flight-certbot/config/software/flight-certbot.rb (hunk 2 above)
Comment (benarmston) [reply builders/flight-certbot/config/software/flight-certbot.rb]: Regarding the symlinks.  Good catch on the `flight_ROOT` setup issue.

However, we now have `/opt/flight/bin/python` and `/opt/flight/opt/python/bin/python` behaving differently.  The first performs the standard flight setup routines and the latter does not.  I wonder if this is going to surprise us later.  Did you consider this difference when you made the change?

File: builders/flight-python/config/software/python.rb (hunk 9 above)
Comment (benarmston) [reply builders/flight-python/config/software/python.rb:66]: That does seem very annoying.  I can't find a way to get that to work (fail) as it should either.

Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/python.rb:66]: ¯\_(ツ)_/¯

File: builders/flight-python/config/software/flight-python.rb (hunk 4 above)
Comment (WilliamMcCumstie) [reply builders/flight-python/config/software/flight-python.rb:73]: > I had been trying to avoid using "conflict" for exactly this reason; it isn't well-defined in this scenario. Hence my requests for you to define it.

Fair point, `conflict` does have a narrowly defined meaning within the context of the builder. Will try and avoid using it in the future.

> Could you clarify that I've understood you correctly

That sums it up quite nicely :+1:

File: builders/flight-certbot/config/software/flight-certbot.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply builders/flight-certbot/config/software/flight-certbot.rb]: Yep I noticed that. I first tried symlinking `/o/f/o/python/bin` to `/o/f/bin` however it broke due to the before mentioned `flight_ROOT` issue. Turns out `BASH_SOURCE` reverse-follows symlinks, TIL.

So the alternative is using `exec`:

1. `/o/f/o/python/bin` first `exec` into `/o/f/bin`,
2. `/of/o/bin` does the regular setup and `exec` into `/o/f/o/python/embedded/bin`

I found this to be counter intuitive as well, so was left with a choice:

1. Jump in and out of `/o/f/o/python` and get the special handling provided by `/o/f/bin`, OR
2. Stay within `/o/f/o/python` and forego the special handling.

---

So the question comes down to: _do we need the special handling?_

AFAICS the special handling does three things:

* Applies "flight" related env vars from a `flight.rc`,
* Applies a bunch of hooks, and
* Ensure `flight_ROOT` is set.

I can't think of any "flight env vars" which will be particularly important for `python` (barring `flight_ROOT`), so the `flight.rc` is probably irrelevant. The same applies to the hooks.

However changing the `flight_ROOT` would make `/of/o/python/bin` execute something other than `/o/f/o/python/embedded/bin`. This would be highly surprising and should be avoided.

---

tl;dr It is probably fine as it, both methods have draw backs.

If you want the standard behaviour use `/o/f/bin`. If you want to guarantee the python you use, select `/o/f/o/python/bin`.

File: builders/flight-python/config/software/flight-python.rb (hunk 4 above)
Comment (benarmston) [reply builders/flight-python/config/software/flight-python.rb:73]: :+1: 

The mechanism implemented seems like a reasonable way of detecting and avoiding this.

File: builders/flight-certbot/config/software/flight-certbot.rb (hunk 2 above)
Comment (benarmston) [reply builders/flight-certbot/config/software/flight-certbot.rb]: I'm not sure if this is the behaviour we want or not.  It is at least consistent with how `flight-nodejs` does things and that seems to work.  If it turns out to be wrong, we'll find out and fix it then.

File: builders/flight-python/config/software/python.rb (hunk 5 above)
Comment (benarmston) [reply builders/flight-python/config/software/python.rb:27]: These are similar enough that its probably fine.

**Comment by benarmston** [comment]
Thanks for your work on this @WilliamMcCumstie.  LGTM.
